testing_system/code_extractor/code_extractor.py

```diff
@@ -0,0 +1,295 @@
+        return [mi.model_name for mi in self.model_info]
+
+    def get_preprocessing_code(self) -> str:
+        """
+        By default returns code above constructor of first encountered model
+        """
+        first_model = self.model_info[0]
+        first_model_cell_no, first_model_line_no = first_model.constructor_location.cell_no, \
+                                                   first_model.constructor_location.line_no - 1
+
+        return NtbCodeExtractorHelper.generate_code_up_to(first_model_cell_no,
+                                                          first_model_line_no,
+                                                          path_to_ntb_file=self.path_to_ntb_file,
+                                                          including_last_line=False)
+
+    def get_preprocessing_script(self) -> str:
+        first_model = self.model_info[0]
+
+        res = ''
+        res += self.get_preprocessing_code()
+        res += self.__generate_preproc_return_stmt(first_model.X_train_var_name, first_model.y_train_var_name)
+
+        return res
+
+    def generate_training_script(self, model_name=None, how_to_pickle='pickle') -> str:
+        if model_name is not None:
+            encountered_model_names = self.get_model_names()
+            if model_name not in encountered_model_names:
+                raise RuntimeError(f'Model {model_name} was not found in {self.path_to_ntb_file} notebook.')
+            if model_name != encountered_model_names[0]:
+                raise RuntimeError(f'Not Implemented. Required model {model_name} is not the first to occur'
+                                   f'in {self.path_to_ntb_file} notebook.')
+
+        if how_to_pickle not in ['pickle', 'joblib']:
+            raise RuntimeError(f'Invalid argument for pickling strategy: {how_to_pickle}')
+
+        first_model = self.model_info[0]
+        if first_model.fit_location is None:
+            raise RuntimeError(f'Fit location is missing for model {model_name}'
+                               f' in {self.path_to_ntb_file} notebook')
+        first_model_fit_cell_no, first_model_fit_line_no = first_model.fit_location.cell_no, \
+                                                           first_model.fit_location.line_no - 1
+        res = ''
+        res += NtbCodeExtractorHelper.generate_code_up_to(first_model_fit_cell_no,
+                                                          first_model_fit_line_no,
+                                                          path_to_ntb_file=self.path_to_ntb_file,
+                                                          including_last_line=True) + '\n'
+
+        model_meta_inf_str = f'{{"training_sample_size": len({first_model.X_train_var_name})}}'
+        res += self.__generate_training_return_stmt(first_model.variable_name,
+                                                    how_to_pickle,
+                                                    meta_inf_as_str=model_meta_inf_str)
```

Comment:
here you have duplicated code - could be simplified but firstly we need to discuss class design

Reply:
I don't see here a lot code duplicates, can you please clarify that point?